File: MageTests/TestingAppDelegate.m
//
//  TestingAppDelegate.m
//  MAGE
//
//  Created by Dan Barela on 2/5/18.
//  Copyright © 2018 National Geospatial Intelligence Agency. All rights reserved.
//

#import "TestingAppDelegate.h"
#import "MagicalRecord+MAGE.h"
#import "GPKGGeoPackageManager.h"
#import "GPKGGeoPackageFactory.h"
#import "MageInitializer.h"

@interface TestingAppDelegate ()
@property (nonatomic, strong) BaseMapOverlay *backgroundOverlay;
@property (nonatomic, strong) BaseMapOverlay *darkBackgroundOverlay;
@end

@implementation TestingAppDelegate

- (void)applicationDidBecomeActive:(UIApplication *)application {
    
}

- (void)applicationWillEnterForeground:(UIApplication *)application {
    
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
    
}

- (void)applicationWillResignActive:(UIApplication *)application {
    
}

- (void)applicationProtectedDataDidBecomeAvailable:(UIApplication *)application {
    
}

- (void)applicationWillTerminate:(UIApplication *)application {
    
}

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    [MageInitializer initializePreferences];
    
    [MagicalRecord setupCoreDataStackWithInMemoryStore];
    [MagicalRecord setLoggingLevel:MagicalRecordLoggingLevelVerbose];
    
    NSString *countriesDarkGeoPackagePath = [[NSBundle mainBundle] pathForResource:@"countries_dark" ofType:@"gpkg"];
    NSLog(@"Countries GeoPackage path %@", countriesDarkGeoPackagePath);
    
    // Add the GeoPackage caches
    GPKGGeoPackageManager * manager = [GPKGGeoPackageFactory manager];
    @try {
        [manager importGeoPackageFromPath:countriesDarkGeoPackagePath];
    }
    @catch (NSException *e) {
        // probably was already imported and that is fine
    }
    NSString *countriesGeoPackagePath = [[NSBundle mainBundle] pathForResource:@"countries" ofType:@"gpkg"];
    NSLog(@"Countries GeoPackage path %@", countriesGeoPackagePath);
    @try {
        [manager importGeoPackageFromPath:countriesGeoPackagePath];
    }
    @catch (NSException *e) {
        // probably was already imported and that is fine
    }
    
    GPKGGeoPackage *countriesGeoPackage = [manager open:@"countries"];
    if (countriesGeoPackage) {
        GPKGFeatureDao * featureDao = [countriesGeoPackage featureDaoWithTableName:@"countries"];
        
        // If indexed, add as a tile overlay
        GPKGFeatureTiles * featureTiles = [[GPKGFeatureTiles alloc] initWithGeoPackage:countriesGeoPackage andFeatureDao:featureDao];
        [featureTiles setIndexManager:[[GPKGFeatureIndexManager alloc] initWithGeoPackage:countriesGeoPackage andFeatureDao:featureDao]];
        
        self.backgroundOverlay = [[BaseMapOverlay alloc] initWithFeatureTiles:featureTiles];
        [self.backgroundOverlay setMinZoom:0];
        self.backgroundOverlay.darkTheme = NO;
        
        self.backgroundOverlay.canReplaceMapContent = true;
    }
    
    GPKGGeoPackage *darkCountriesGeoPackage = [manager open:@"countries_dark"];
    if (darkCountriesGeoPackage) {
        GPKGFeatureDao * darkFeatureDao = [darkCountriesGeoPackage featureDaoWithTableName:@"countries"];
        
        // If indexed, add as a tile overlay
        GPKGFeatureTiles * darkFeatureTiles = [[GPKGFeatureTiles alloc] initWithGeoPackage:darkCountriesGeoPackage andFeatureDao:darkFeatureDao];
        [darkFeatureTiles setIndexManager:[[GPKGFeatureIndexManager alloc] initWithGeoPackage:darkCountriesGeoPackage andFeatureDao:darkFeatureDao]];
        
        self.darkBackgroundOverlay = [[BaseMapOverlay alloc] initWithFeatureTiles:darkFeatureTiles];
        [self.darkBackgroundOverlay setMinZoom:0];
        self.darkBackgroundOverlay.darkTheme = YES;
        
        self.darkBackgroundOverlay.canReplaceMapContent = true;
    }
    
    return YES;
}

- (void) logout {
    self.logoutCalled = YES;
}

- (BaseMapOverlay *) getBaseMap {
    return self.backgroundOverlay;
}

- (BaseMapOverlay *) getDarkBaseMap {
    return self.darkBackgroundOverlay;
}

@end

File: Mage/TestingAppDelegate.swift
//
//  TestingAppDelegate.swift
//  MAGE
//
//  Created by Brent Michalski on 9/28/25.
//  Copyright © 2025 National Geospatial Intelligence Agency. All rights reserved.
//

import UIKit
import GeoPackage

@objc(TestingAppDelegate)
class TestingAppDelegate: AppDelegate {
    @objc dynamic var logoutCalled = false
    
    private var backgroundOverlay: BaseMapOverlay?
    private var darkBackgroundOverlay: BaseMapOverlay?
    
    override func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil
    ) -> Bool {
        MageInitializer.initializePreferences()
        backgroundOverlay = MageInitializer.getBaseMap()
        darkBackgroundOverlay = MageInitializer.getDarkBaseMap()
        
        // Ensure a key window exists for KIF/UIKit assumptions
        let win = UIWindow(frame: UIScreen.main.bounds)
        win.rootViewController = UIViewController()
        win.makeKeyAndVisible()
        self.window = win
        
        return true
    }
    
    // No-ops to match the old Obj-C stubs (kept for parity)
    override func applicationDidBecomeActive(_ application: UIApplication) {}
    override func applicationWillEnterForeground(_ application: UIApplication) {}
    override func applicationDidEnterBackground(_ application: UIApplication) {}
    override func applicationWillResignActive(_ application: UIApplication) {}
    override func applicationProtectedDataDidBecomeAvailable(_ application: UIApplication) {}
    override func applicationWillTerminate(_ application: UIApplication) {}
    
    // Test hooks
    @objc override func logout() { logoutCalled = true }
    @objc override func getBaseMap() -> BaseMapOverlay! { MageInitializer.getBaseMap() }
    @objc override func getDarkBaseMap() -> BaseMapOverlay! { MageInitializer.getDarkBaseMap() }
}

@objc(ViewLoaderAppDelegate)
final class ViewLoaderAppDelegate: UIResponder, UIApplicationDelegate {
    var window: UIWindow?
    func application(_ application: UIApplication,
                     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil) -> Bool {
        let win = UIWindow(frame: UIScreen.main.bounds)
        // Provide a simple view host if you need it
        win.rootViewController = UIViewController()
        win.makeKeyAndVisible()
        window = win
        return true
    }
}
